zeppelin-zengine/src/main/java/org/apache/zeppelin/scheduler/pool/DynamicThreadPool.java

```diff
@@ -0,0 +1,248 @@
+   * shutdown in which previously submitted tasks are executed, but no new tasks will be accepted.
+   * @see ExecutorService#shutdown() if false: Attempts to stop all actively executing tasks, halts
+   * the processing of waiting tasks.
+   * @see ExecutorService#shutdownNow()
+   */
+  @Override
+  public void shutdown(boolean waitForJobsToComplete) {
+    LOGGER.info("Shutting down ThreadPool...");
+    if (!waitForJobsToComplete) {
+      try {
+        executor.shutdown();
+        final boolean done = executor.awaitTermination(1, TimeUnit.MINUTES);
+        LOGGER.info("Have all the cron tasks been completed? {}", done);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+      }
+    } else {
+      executor.shutdownNow();
+      final List<Runnable> rejected = executor.shutdownNow();
+      // Should be zero because of SynchronousQueue
+      LOGGER.info("Rejected tasks: {}: {}", rejected.size(), rejected);
+    }
+    INSTANCES.remove(schedulerInstanceId);
+  }
+
+  public void setThreadNamePrefix(String prfx) {
+    this.threadNamePrefix = prfx;
+  }
+
+  /**
+   * Set the number of worker threads in the pool. Increases <code>{@link
+   * ThreadPoolExecutor#maximumPoolSize}</code>
+   */
+  public void setThreadCount(int threadCount) throws SchedulerConfigException {
+    if (executor.getMaximumPoolSize() == threadCount) {
+      return;
+    }
+    LOGGER.info(
+        "Changing ThreadPool size from {} to {}", executor.getMaximumPoolSize(), threadCount);
+    try {
+      executor.setMaximumPoolSize(threadCount);
+    } catch (IllegalArgumentException e) {
+      throw new SchedulerConfigException("Thread count must be > 0");
+    }
+  }
+
+  public void setKeepAliveTimeSec(long time) {
+    executor.setKeepAliveTime(time, TimeUnit.SECONDS);
+  }
+
+  /**
+   * Get the maximum number of worker threads in the pool.
+   */
+  @Override
+  public int getPoolSize() {
+    return executor.getMaximumPoolSize();
+  }
+
+  @Override
+  public void setInstanceId(String schedInstId) {
+    if (schedulerInstanceId != null) {
+      INSTANCES.remove(schedulerInstanceId, this);
+    }
+    schedulerInstanceId = schedInstId;
+    if (schedulerInstanceId != null) {
+      INSTANCES.put(schedulerInstanceId, this);
+    }
+  }
+
+  @Override
+  public void setInstanceName(String schedName) {
+    this.schedulerInstanceName = schedName;
+  }
+
+  /**
+   * Set the thread priority of worker threads in the pool.
+   */
+  public void setThreadPriority(int priority) {
+    this.priority = priority;
+  }
+
+  private class DynamicThreadFactory implements ThreadFactory {
+    private final AtomicLong count = new AtomicLong(0L);
+
+    @Override
+    public Thread newThread(Runnable runnable) {
+      Thread newThread = new Thread(runnable);
+      newThread.setUncaughtExceptionHandler((th, e) -> LOGGER.error("ThreadPool error", e));
```

Comment:
Я бы сказал, что это 
```suggestion
      newThread.setUncaughtExceptionHandler((th, e) -> LOGGER.error("Unexpected cron job error", e));
```
у пула-то всё хорошо. А вот job упал.